### PR TITLE
Refactor native build to remove the compiler-override files

### PR DESCRIPTION
**Pull request by jkoritzinsky**
- Refactor the native build to remove the compiler-override files
- Use CMake 3.0 on our systems that have 2.X and 3.0 installed.
- Set `CMAKE_USER_MAKE_RULES_OVERRIDE` to an empty string to update the CMake cache so CMake doesn't go looking for the compiler-override files.

### CODE REVIEW

**Comment by jkoritzinsky** [comment]
@janvorli @am11 can you take a look when you have a chance?

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,12 +1,18 @@
 # Verify minimum required version
 cmake_minimum_required(VERSION 2.8.12)
 
-if(CMAKE_VERSION VERSION_EQUAL 3.0 OR CMAKE_VERSION VERSION_GREATER 3.0)
+if(${CMAKE_VERSION} VERSION_EQUAL 3.0 OR ${CMAKE_VERSION} VERSION_GREATER 3.0)
```
Comment (janvorli) [inline CMakeLists.txt]: The ${} you've added should not be necessary here.

File: src/ToolBox/SOS/lldbplugin/CMakeLists.txt [hunk 2]
```diff
@@ -124,7 +124,7 @@ endif()
 
 message(STATUS "LLDB_H: ${LLDB_H}")
 
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-delete-non-virtual-dtor")
+add_compile_options(-Wno-delete-non-virtual-dtor)
```
Comment (janvorli) [inline src/ToolBox/SOS/lldbplugin/CMakeLists.txt]: This (and other similar places) essentially reverts a change that was made intentionally for compilation with GCC in #22687.

Comment (jkoritzinsky) [reply src/ToolBox/SOS/lldbplugin/CMakeLists.txt]: What's the specific reasoning that we need to use `CMAKE_CXX_FLAGS` instead of `add_compile_options`? (for example we need to be able to remove these options in lower directories) I'm fine switching it back, but I want to make sure we have a reason why we're being inconsistent.

File: src/pal/tools/gen-buildsys-clang.sh [hunk 3]
```diff
@@ -147,19 +147,14 @@ if [ "$build_arch" == "armel" ]; then
...
-      (  "$build_arch" == "arm" || "$build_arch" == "armel" ) ]]; then
-    overridefile=clang-compiler-override-arm.txt
-else
-    overridefile=clang-compiler-override.txt
-fi
-
 __currentScriptDir="$script_dir"
 
-cmake_command=cmake
+if command -v "cmake3" > /dev/null 2>&1
```
Comment (janvorli) [inline src/pal/tools/gen-buildsys-clang.sh]: A nit - this can be simplified to cmake_command=$(command -v cmake3 || command -v cmake)

File: src/ToolBox/SOS/lldbplugin/CMakeLists.txt (hunk 2 above)
Comment (janvorli) [reply src/ToolBox/SOS/lldbplugin/CMakeLists.txt]: I believe the reason was that these options apply only to C++ compiler and the gcc C compiler complains or fails on those. @franksinankaya am I correct that that was the reason why you've made this change?

Comment (jkoritzinsky) [reply src/ToolBox/SOS/lldbplugin/CMakeLists.txt]: I just saw the reasoning (GCC for C specifically gets unhappy with some of the warning flags). I'll wait on this portion of the cleanup until we move to a version of CMake with the `$<COMPILE_LANGUAGE>` generator expression.

Comment (franksinankaya) [reply src/ToolBox/SOS/lldbplugin/CMakeLists.txt]: Reasoning is correct.

File: configurecompiler.cmake [hunk 4]
```diff
@@ -184,7 +184,10 @@ endif()
 if(WIN32)
     add_compile_options(/Zi /FC /Zc:strictStrings)
 elseif (CLR_CMAKE_PLATFORM_UNIX)
-    add_compile_options(-g -Wall)
+    add_compile_options(-g)
+    # We need to add -Wall to CMAKE_CXX_FLAGS since add_comiple_options takes precedence
```
Comment (franksinankaya) [inline configurecompiler.cmake]: typo comiple_options

**Comment by franksinankaya** [comment]
I think I want to do a build with gcc on this pull request when it comes to finalizing stage. Please let me know.

**Comment by jkoritzinsky** [comment]
@franksinankaya can you do a build with GCC?

**Comment by franksinankaya** [comment]
96% compiled successfully. 

"-Werror=conversion-null" doesn't seem to propagate to for some reason.

src/ToolBox/SOS/Strike/strike.cpp:14788:24: error: converting to non-pointer type ‘TADDR {aka long unsigned int}’ from NULL [-Werror=conversion-null]
         TADDR taLTOH = NULL;


You can clone this repo and build with ./build.sh -gcc

https://github.com/franksinankaya/coreclr/tree/cmake-cleanup

to reproduce.

**Comment by jkoritzinsky** [comment]
@franksinankaya looks like that one was another precedence issue between CMAKE_CXX_FLAGS and add_compile_options. I've pushed out a fix. Let me know if it works for you.

**Comment by franksinankaya** [comment]
Thanks this version worked.

**Comment by jkoritzinsky** [comment]
@janvorli can you take another review pass?

File: tests/CMakeLists.txt [hunk 5]
```diff
@@ -1,15 +1,22 @@
 # Require at least version 2.8.12 of CMake
 cmake_minimum_required(VERSION 2.8.12)
 
-# Include global configure settings
-include(${CMAKE_CURRENT_SOURCE_DIR}/../configure.cmake)
-include(${CMAKE_CURRENT_SOURCE_DIR}/../configurecompiler.cmake)
+if (NOT WIN32)
+  set(CMAKE_CXX_FLAGS_INIT "${CMAKE_CXX_FLAGS_INIT} -std=c++11"  CACHE STRING "Flags used by the compiler during all build types.")
+  set(CMAKE_C_FLAGS_INIT "${CMAKE_C_FLAGS_INIT} -std=c11"  CACHE STRING "Flags used by the compiler during all build types.")
```
Comment (am11) [inline tests/CMakeLists.txt]: The only C11 feature used in CoreCLR is `_Static_assert` in `src/corefx/System.Globalization.Native/pal_compiler.h`. The rest of libunwind and Globalization is gnu99 compatible (c99 plus GNU extensions: named variadic macros and statement expressions etc.). However, `-pedantic`ally, I think we need `-std=gnu11 -pedantic -Wno-gnu -Wno-variadic-macros -Wno-format-pedantic -Wno-pointer-arith -Wno-keyword-macro` for C in CoreCLR.

CoreFX uses gnu99 (the initial reason was: mono compatibility) but the code has C11's `_Generic` as well as `_Static_assert`, which makes it non-conforming to the compiler's `-std` filag. IMO, we should select the minimal standard that is actually used by the code (`-pedantic` highlights where we violate the conformance).

Comment (janvorli) [reply tests/CMakeLists.txt]: The `_Static_assert` is optional in both coreclr and corefx (we check if it exists and if not, we use another way). As for the `_Generic` in corefx, you are right.
I would not change this stuff in this change though.

File: configureoptimization.cmake [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+elseif(CLR_CMAKE_PLATFORM_UNIX)
+    set(CLR_CMAKE_ARM_OPTIMIZATION_FALLBACK OFF)
+    if(CLR_CMAKE_TARGET_ARCH STREQUAL "arm" OR CLR_CMAKE_TARGET_ARCH STREQUAL "armel")
+        if(("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang") AND (CMAKE_CXX_COMPILER_VERSION VERSION_LESS 3.9))
+            set(CLR_CMAKE_ARM_OPTIMIZATION_FALLBACK ON)
+        endif()
+    endif()
+    add_compile_options($<$<CONFIG:Debug>:-O0>)
+    if (CLR_CMAKE_ARM_OPTIMIZATION_FALLBACK)
+        add_compile_options($<$<NOT:$<CONFIG:Debug>>:-O1>)
```
Comment (janvorli) [inline configureoptimization.cmake]: This is not correct. It should be changing the Checked, Release and RelWithDebInfo configurations, not the Debug one.
I am actually surprised the conditional expressions work for Unix generators, we use them only for the multi-generator (for Windows) at other places.

File: src/jit/CMakeLists.txt [hunk 7]
```diff
@@ -1,5 +1,4 @@
 set(CMAKE_INCLUDE_CURRENT_DIR ON)
-set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
```
Comment (janvorli) [inline src/jit/CMakeLists.txt]: Why was this removed?

File: src/pal/tools/gen-buildsys-clang.sh [hunk 8]
```diff
@@ -168,17 +158,18 @@ if [[ "$scan_build" == "ON" ]]; then
...
+# Include CMAKE_USER_MAKE_RULES_OVERRIDE as uninitialized since it will hold its value in the CMake cache otherwise can cause issues when branch switching
 $cmake_command \
   -G "$generator" \
-  "-DCMAKE_USER_MAKE_RULES_OVERRIDE=${__currentScriptDir}/$overridefile" \
   "-DCMAKE_AR=$llvm_ar" \
   "-DCMAKE_LINKER=$llvm_link" \
   "-DCMAKE_NM=$llvm_nm" \
   "-DCMAKE_OBJDUMP=$llvm_objdump" \
   "-DCMAKE_BUILD_TYPE=$buildtype" \
-  "-DCMAKE_EXPORT_COMPILE_COMMANDS=1 " \
```
Comment (janvorli) [inline src/pal/tools/gen-buildsys-clang.sh]: Why was this removed?

File: tests/CMakeLists.txt [hunk 9]
```diff
@@ -1,15 +1,22 @@
...
+endif()
+
+project(Project)
```
Comment (janvorli) [inline tests/CMakeLists.txt]: Was adding this necessary?

File: tests/CMakeLists.txt [hunk 10]
```diff
@@ -1,15 +1,22 @@
...
 
 set(INC_PLATFORM_DIR ${CMAKE_CURRENT_SOURCE_DIR}/src/Common/Platform)
 if (WIN32)
     add_definitions(-DWINDOWS=1)
 endif()
 
+# Include global configure settings
+include(${CMAKE_CURRENT_SOURCE_DIR}/../configure.cmake)
+include(${CMAKE_CURRENT_SOURCE_DIR}/../configurecompiler.cmake)
+include(${CMAKE_CURRENT_SOURCE_DIR}/../configureoptimization.cmake)
```
Comment (janvorli) [inline tests/CMakeLists.txt]: It seems that configuring optimizations is part of configuring the compiler. I would suggest either include the `configureoptimization.cmake` from the `configurecompiler.cmake` or move the stuff from `configureoptimization.cmake` into the `configurecompiler.cmake`

File: src/jit/CMakeLists.txt (hunk 7 above)
Comment (jkoritzinsky) [reply src/jit/CMakeLists.txt]: This is done at the root CMakeLists.txt now.

File: tests/CMakeLists.txt (hunk 9 above)
Comment (jkoritzinsky) [reply tests/CMakeLists.txt]: For the `*_INIT` properties to work correctly, the languages need to be enabled after setting them. I can try some variations here (using just the `CMAKE_<LANG>_FLAGS` properties instead of the init variants or using `enable_language()`) if we don't want to have the `project()` call.

File: configureoptimization.cmake (hunk 6 above)
Comment (jkoritzinsky) [reply configureoptimization.cmake]: This is changing all of the non-Debug configurations. The generator expressions work for all CMake generators, but for non-multi-generators, you specify the configuration with `CMAKE_BUILD_TYPE` (which is what we do).

File: src/pal/tools/gen-buildsys-clang.sh (hunk 8 above)
Comment (jkoritzinsky) [reply src/pal/tools/gen-buildsys-clang.sh]: The goal here was to centralize where we set `CMAKE_EXPORT_COMPILE_COMMANDS` to a single location in the tree (as part of limiting the number of defines we have to pass to CMake on the command line). I put it in the root CMakeLists.txt file. I'll can move it to configurecompiler.cmake if we want the compilation database generated for the test build as well.

File: tests/CMakeLists.txt (hunk 5 above)
Comment (am11) [reply tests/CMakeLists.txt]: Sure, btw I have this fixed in ongoing solaris port branch, I will rebase once this PR is merged.
gcc 7 on SmartOS amd64 throws the following error on building Globalization target with `-std=c11`, with gets fixed with `-std=gnu99` or `-std=gnu11` flag (without changing the code or other flags):

```sh
/datadrive/projects/coreclr/src/corefx/System.Globalization.Native/pal_calendarData.c:84:9: error: implicit declaration of function 'strcasecmp'; did you mean 'u_strcasecmp'? [-Werror=implicit-function-declaration]
     if (strcasecmp(calendarName, GREGORIAN_NAME) == 0)
         ^~~~~~~~~~
         u_strcasecmp
```

File: configureoptimization.cmake (hunk 6 above)
Comment (janvorli) [reply configureoptimization.cmake]: Oops,  I am sorry, my eyes have missed the `NOT` in the condition.

File: tests/CMakeLists.txt (hunk 5 above)
Comment (jkoritzinsky) [reply tests/CMakeLists.txt]: That warning I would think is actually a problem you should look at fixing (the declaration of `strcasecmp` isn't found anywhere) even though the "implicit definition" behavior happens to match the signature.

**Comment by jkoritzinsky** [comment]
@janvorli any more feedback on this PR?
